articles/modules/ROOT/pages/using-subqueries-to-avoid-the-eager.adoc

```diff
@@ -0,0 +1,222 @@
+
+Batch commits are not compatible with Eager behavior, as they require lazy row-by-row semantics for correct operation.
+While Cypher does not stop you from attempting batch commit operations when there is an Eager in the plan, they will not commit in batches, and you may encounter the above mentioned issues around GC pauses and heap problems.
+
+Why does this happen?
+
+Cypher semantics demand that to the greatest extent possible, operations from later in the query should not influence the results of operations from earlier in the query.
+For example, a MERGE that appears later in the query should not influence a MATCH that shows up earlier in the query (on nodes of the same label).
+
+At first glance such a rule may seem nonsensical, especially if only considering a single row of input, since of course a MATCH operation earlier in the query would execute before a MERGE that happens later in the query.
+
+However, the problem of ordering and influencing of results makes more sense when considering queries that execute over multiple rows, either from pure input, such as LOAD CSV, or from MATCH operations that can return many rows.
+Cypher planning would ordinarily try for row-by-row processing, so the entire remaining query would execute for each input row.
+Because of that, a MERGE from later in the query, applying to an earlier row of input, would happen before processing of a later row of input, that has yet to execute its MATCH operation.
+
+If that MERGE from later in the query could affact the results of a MATCH earlier in the query, then that violates the above mentioned Cypher semantics, and so Eager is planned to preserve them.
+This causes the change in execution behavior, so instead of lazy row-by-row processing, all rows are processed operation-by-operation.
+
+If the input size is too large, either from the very start, or building  up over the course of execution, since all interim results must be built up and held in memory at once for all rows, this could easily exceed the bounds of the heap and cause out of memory errors, thus the problem of Eager.
+
+We have a blog entry by Jennifer Reif discussing Eager and its effects in more detail here:
+
+https://community.neo4j.com/t5/general-discussions/cypher-sleuthing-the-eager-operator/m-p/50596
+
+Operations that are likely to result in Eager being planned include these in the query, when there is are operations preceding them (MATCH, MERGE, UNWIND, CALL, or LOAD CSV) that suggest multiple input rows to process:
+
+* MATCH (regular or OPTIONAL) and CREATE clauses (in any ordering) on the same node labels
+* MATCH (regular or OPTIONAL) and MERGE clauses (in any ordering) on the same node labels
+* CREATE and MERGE clauses (in any ordering) on the same node labels
+* Multiple MERGE clauses on the same labels
```

Comment:
which can happen if you load a mono-partite graph like `(:User)-[:FOLLOWS]->(:User)`